### DISCUSSION TITLE
how to access private devops repo

### DISCUSSION THREAD

**Discussion by marvinfischenich** [renovatebot/renovate#16279] [Request Help]
Hi guys,
i'm a bit confused since i want to add a private devops repo to my hostrule but nothing i tried worked so far.

I have tried several combinations but nothing worked so far ... thought that couldn't be that complicated.
Hoping for you help!

```
  "hostRules": [
    {
      "matchHost": "https:/dev.azure.com/myorg",
      "token": "abc123"
    },
{
      "matchHost": "https:/dev.azure.com/myorg/project",
      "token": "abc123"
    },
{
      "matchHost": "https:/dev.azure.com/myorg/project/_git",
      "token": "abc123"
    },
{
      "matchHost": "https:/dev.azure.com/myorg/project/_git/repo1",
      "token": "abc123"
    },
{
      "matchHost": "https:/dev.azure.com/myorg/project/_git/repo2",
      "token": "abc123"
    }

  ],
```

The devops repo URL's are for example
https://dev.azure.com/myorg/project/_git/repo1
https://dev.azure.com/myorg/project/_git/repo2

The dry-run pipeline then fails with:
`DEBUG: Datasource 404 (repository=apps/base-image-update/base-image-update)
       "datasource": "pypi",
       "packageName": "repo1",
       "url": "https://pypi.org/project/repo1/"
DEBUG: Failed to look up dependency repo1 (repository=apps/base-image-update/base-image-update, packageFile=requirements.txt, dependency=repo1)`

But somehow it is detecting the repos on devops:

```
DEBUG: Adding token authentication for https://dev.azure.com/myorg/project/_git/repo1 to hostRules (repository=apps/base-image-update/base-image-update)
DEBUG: Adding token authentication for https://dev.azure.com/myorg/project/_git/repo2 to hostRules (repository=apps/base-image-update/base-image-update)
DEBUG: Found repo ignorePaths (repository=apps/base-image-update/base-image-update)
       "ignorePaths": [
         "**/node_modules/**",
         "**/bower_components/**",
         "**/vendor/**",
         "**/examples/**",
         "**/__tests__/**",
         "**/test/**",
         "**/tests/**",
         "**/__fixtures__/**"
       ]
```

So i don't know why it isn't detecting dependencies to repo1 and repo2.

Thans a lot for your help!

**Comment by rarkins** [comment]
Why are you adding hostRules in the first place? Are you just trying to get started?

**Comment by marvinfischenich** [comment]
Yes its my first touch point with renovate bot.
What would be the better alternative to hostRules?


Rhys Arkins ***@***.***> schrieb am Mo., 27. Juni 2022, 21:35:

**Reply by viceice** [reply]
using hostrules is best approach. Please describe in more detail what you're trying to achive.

You can also post your sanitized config heree, so we can have a look

**Reply by rarkins** [reply]
I was wondering if this was their first attempt to set up a single repo. i.e. they just need to specify `token` etc properly

**Reply by marvinfischenich** [reply]
i want to add a private repo from devops as dependency so related repos get automatic updates as dependency repo1 or repo2 are updated. the token has full access.
my config is:
`
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>renovatebot/.github"],
  "includeForks": true,
  "platformAutomerge": true,
  "platform": "gitlab",
  "hostRules": [
    {
      "matchHost": "https:/dev.azure.com/myorg",
      "token": "xky...4ca"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/project",
      "token": "xkyk...4ca"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/project/_git",
      "token": "xky...4ca"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/project/_git/repo1",
      "token": "xky...4ca"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/project/_git/repo2",
      "token": "xky...4ca"
    }
  ],
  "packageRules": [
    {
      "description": "Trigger breaking release for major updates",
      "matchPackageNames": ["docker", "renovate/renovate"],
      "matchUpdateTypes": ["major"],
      "semanticCommitType": "feat",
      "commitBody": "BREAKING CHANGE: Major update",
      "automergeType": "pr"
    },
    {
      "description": "Trigger feature release for minor updates",
      "matchPackageNames": ["docker", "renovate/renovate"],
      "matchUpdateTypes": ["minor"],
      "semanticCommitType": "feat",
      "automergeType": "pr"
    },
    {
      "description": "Trigger fix release for patch updates",
      "matchPackageNames": ["docker", "renovate/renovate"],
      "matchUpdateTypes": ["patch", "digest"],
      "semanticCommitType": "fix",
      "automergeType": "pr"
    },
    {
      "description": "Allow updates after 3 days (exclude renovate)",
      "excludePackageNames": ["renovate/renovate"],
      "separateMinorPatch": true,
      "stabilityDays": 3
    },
    {
      "description": "Schedule digest updates daily",
      "matchUpdateTypes": ["digest"],
      "extends": ["schedule:daily"]
    },
    {
      "matchPackageNames": ["renovate/node"],
      "versioning": "node"
    },
    {
      "description": "Full image is sometimes behind, so we need to force a new MR",
      "matchPackageNames": ["renovate/renovate"],
      "matchUpdateTypes": ["major", "minor", "patch"],
      "recreateClosed": true
    }
  ],
  "regexManagers": [
    {
      "description": "Update docker references in gitlab variables",
      "fileMatch": ["\\.gitlab-ci\\.yml$"],
      "matchStrings": [
        "CI_RENOVATE_IMAGE\\s*:\\s*(?<depName>[^:]+?):(?<currentValue>[a-z0-9.-]+)(?:@(?<currentDigest>sha256:[a-f0-9]+))?",
        "CI_RENOVATE_SERVICE\\s*:\\s*(?<depName>[^:]+?):(?<currentValue>[a-z0-9.-]+)(?:@(?<currentDigest>sha256:[a-f0-9]+))?"
      ],
      "datasourceTemplate": "docker",
      "versioningTemplate": "docker"
    }
  ],
  "gitlabci": {
    "fileMatch": ["\\.gitlab-ci\\.yml$"],
    "pinDigests": true
  }
}
`

**Reply by HonkingGoose** [reply]
I've put your config through Prettier, so we can read it. 😉 

<details><summary>Click me to see readable config</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>renovatebot/.github"],
  "includeForks": true,
  "platformAutomerge": true,
  "platform": "gitlab",
  "hostRules": [
    { "matchHost": "https:/dev.azure.com/myorg", "token": "xky...4ca" },
    {
      "matchHost": "https://dev.azure.com/myorg/project",
      "token": "xkyk...4ca"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/project/_git",
      "token": "xky...4ca"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/project/_git/repo1",
      "token": "xky...4ca"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/project/_git/repo2",
      "token": "xky...4ca"
    }
  ],
  "packageRules": [
    {
      "description": "Trigger breaking release for major updates",
      "matchPackageNames": ["docker", "renovate/renovate"],
      "matchUpdateTypes": ["major"],
      "semanticCommitType": "feat",
      "commitBody": "BREAKING CHANGE: Major update",
      "automergeType": "pr"
    },
    {
      "description": "Trigger feature release for minor updates",
      "matchPackageNames": ["docker", "renovate/renovate"],
      "matchUpdateTypes": ["minor"],
      "semanticCommitType": "feat",
      "automergeType": "pr"
    },
    {
      "description": "Trigger fix release for patch updates",
      "matchPackageNames": ["docker", "renovate/renovate"],
      "matchUpdateTypes": ["patch", "digest"],
      "semanticCommitType": "fix",
      "automergeType": "pr"
    },
    {
      "description": "Allow updates after 3 days (exclude renovate)",
      "excludePackageNames": ["renovate/renovate"],
      "separateMinorPatch": true,
      "stabilityDays": 3
    },
    {
      "description": "Schedule digest updates daily",
      "matchUpdateTypes": ["digest"],
      "extends": ["schedule:daily"]
    },
    { "matchPackageNames": ["renovate/node"], "versioning": "node" },
    {
      "description": "Full image is sometimes behind, so we need to force a new MR",
      "matchPackageNames": ["renovate/renovate"],
      "matchUpdateTypes": ["major", "minor", "patch"],
      "recreateClosed": true
    }
  ],
  "regexManagers": [
    {
      "description": "Update docker references in gitlab variables",
      "fileMatch": ["\\.gitlab-ci\\.yml$"],
      "matchStrings": [
        "CI_RENOVATE_IMAGE\\s*:\\s*(?<depName>[^:]+?):(?<currentValue>[a-z0-9.-]+)(?:@(?<currentDigest>sha256:[a-f0-9]+))?",
        "CI_RENOVATE_SERVICE\\s*:\\s*(?<depName>[^:]+?):(?<currentValue>[a-z0-9.-]+)(?:@(?<currentDigest>sha256:[a-f0-9]+))?"
      ],
      "datasourceTemplate": "docker",
      "versioningTemplate": "docker"
    }
  ],
  "gitlabci": { "fileMatch": ["\\.gitlab-ci\\.yml$"], "pinDigests": true }
}
```
</details>

**Reply by rarkins** [reply]
What type of dependency is it? e.g. can you find it in the `packageFiles with updates` log message?

**Reply by marvinfischenich** [reply]
the dependency is detected correctly.
however, renovate doesn't use my private devops to verify the newest version but public pypi instead.
there it won't find the dependency since it's private:
`
DEBUG: packageFiles with updates (repository=apps/base-image-update/base-image-update)

       "config": {
         "gitlabci-include": [
           {
             "packageFile": ".gitlab-ci.yml",
             "deps": [
               {
                 "datasource": "gitlab-tags",
                 "depName": "gitlabci/app",
                 "depType": "repository",
                 "skipReason": "unknown-version",
                 "registryUrls": ["https://git.myservice.net"],
                 "depIndex": 0,
                 "updates": []
               }
             ]
           }
         ],
         "pip_requirements": [
           {
             "packageFile": "requirements.txt",
             "deps": [
               {
                 "depName": "repo1",
                 "currentValue": "== 1.5.26",
                 "datasource": "pypi",
                 "currentVersion": "1.5.26",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "repo1",
                     "message": "Failed to look up dependency repo1"
                   }
                 ],
                 "versioning": "pep440"
               }
             ]
           }
         ]
       }
`

**Reply by rarkins** [reply]
You're showing two dependencies in that log. Are you referring to the second one?

**Reply by marvinfischenich** [reply]
yes i'm referring to the 2nd one

**Reply by rarkins** [reply]
So the first step is that Renovate needs to know *where* to look it up. As you saw, it's looking it up from PyPI and not your custom registry. To teach Renovate where to find the dependency, you need to create a `packageRules` entry with matchDatasources=pypi, matchPackageNames=repo1, and registryUrls=["https://azure-pypi-registry-url"]. Only after that works will Renovate know to try your custom registry, and and then you can use hostRules to add credentials.

**Reply by marvinfischenich** [reply]
now it looks better thanks, thought renovate is going to do that automatically

**Reply by rarkins** [reply]
How could it learn or guess which registry to use for that dependency? Eg is there something in your requirements.txt ?

**Reply by marvinfischenich** [reply]
yes, the dependency is mentioned in the requirements.txt.
i thought it would just search all available public registry and look for the dependency.

**Reply by rarkins** [reply]
Can you provide an example `requirements.txt`?

**Reply by marvinfischenich** [reply]
that's it, nothing special i assume.

[requirements.txt](https://github.com/renovatebot/renovate/files/9052743/requirements.txt)

**Reply by rarkins** [reply]
There is no **registry** in that file, only a dependency name. So how could Renovate learn or guess which registry to use?

**Reply by marvinfischenich** [reply]
the registry is defined in the same repo in the renovate.json. isn't that the proper place?

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>renovatebot/.github"],
  "includeForks": true,
  "platformAutomerge": true,
  "platform": "gitlab",
  "hostRules": [
    {
      "matchHost": "https://dev.azure.com/myorg/repo1/",
      "token": "..."
    },
    {
      "matchHost": "https://dev.azure.com/myorg/repo1/_git/",
      "token": "..."
    }    
  ],
  "packageRules": [
    {
      "matchDatasources":"pypi",
      "matchPackagePatterns": ["^repo1-"],
      "registryUrls":[
        "https://dev.azure.com/myorg/repo1",
        "https://dev.azure.com/myorg/repo1/_git/"
      ]
    }
  ]
}

```

where can i find it in the docs how to setup the registry in the requirements.txt?

**Reply by rarkins** [reply]
When I asked you, "How could it learn or guess which registry to use for that dependency? Eg is there something in your requirements.txt ?" I meant is the registry in the `requirements.txt`. You answered Yes
